File: typescript/src/tagged-tuple.ts
import { bencoder } from "./bencode.ts";
import type { Encoding } from "node:crypto";

type Tagged<T, A extends unknown[], Tag extends number> = {
  new(...args: A): T;
  readonly tag: Tag;
};

export abstract class TaggedTuple {
  public static fromTuple<
    T extends TaggedTuple,
    A extends unknown[],
    Tag extends number,
  >(this: Tagged<T, A, Tag>, tag: NoInfer<Tag>, ...tuple: A): T {
    if (tag !== this.tag) {
      throw new Error(`Tag mismatch: expected ${this.tag}, got ${tag}`);
    }
    return new this(...tuple);
  }

  public asTuple(): [number, ...unknown[]] {
    const { tag } = this.constructor as typeof TaggedTuple & { tag: number };
    return [tag, ...Object.values(this)];
  }
}

export abstract class TaggedTupleStrings extends TaggedTuple {
  private static readonly encoding: Encoding = "utf-8";

  public encode(): Uint8Array {
    const tuple = this.asTuple();
    return bencoder.encode(tuple);
  }

  public static decode<
    T extends TaggedTuple,
    A extends unknown[],
    Tag extends number,
  >(this: Tagged<T, A, Tag>, data: Uint8Array): T {
    const decoded = bencoder.decode(data, TaggedTupleStrings.encoding) as [
      Tag,
      ...A,
    ];
    // cheating here - but we have enough type and test coverage to be confident
    return (this as any).fromTuple(...decoded);
  }
}

export class PendingReturn extends TaggedTuple {
  public static readonly tag = 1;

  public readonly root_id: string;
  public readonly call_hash: string;
  public readonly topic: string;

  constructor(root_id: string, call_hash: string, topic: string) {
    super();
    this.root_id = root_id;
    this.call_hash = call_hash;
    this.topic = topic;
  }
}

export class ScheduleMessage extends TaggedTupleStrings {
  public static readonly tag = 2;

  public readonly root_id: string;
  public readonly call_hash: string;

  constructor(root_id: string, call_hash: string) {
    super();
    this.root_id = root_id;
    this.call_hash = call_hash;
  }
}

File: python/src/brrr/tagged_tuple.py
import dataclasses
from dataclasses import dataclass
from typing import Any, ClassVar, Self

import bencodepy

_bc = bencodepy.Bencode(encoding="utf-8")


@dataclass(frozen=True)
class TaggedTuple:
    """Ad-hoc static registry of record serialization semantics.

    Compare to protobuf but only semantically (the actual encoding/decoding is
    expected to be done by bencode) and far less featureful.

    The responsibility is on the developer to never re-use a version and always
    bump the version number when anything changes.

    The version is just meant as a stateless, light-weight encoding/decoding
    assert--not for any runtime type inference.

    The semantics are light weight and can be encoded using bencode using
    minimal overhead (which is why it's a tuple instead of a dict).

    These datastructures are _not_ part of the public brrr API, but they _are_
    part of the brrr wire protocol, meaning they must be supported by all
    languages implementing brrr.

    """

    tag: ClassVar[int]

    def astuple(self) -> tuple[Any, ...]:
        return (self.tag,) + dataclasses.astuple(self)

    @classmethod
    def fromtuple(cls, t: tuple[Any, ...]) -> Self:
        if t[0] != cls.tag:
            raise ValueError(f"{cls.__name__} decode tag mismatch: {t[0]} != {cls.tag}")
        return cls(*t[1:])


@dataclass(frozen=True)
class TaggedTupleStrings(TaggedTuple):
    def encode(self) -> bytes:
        x: bytes = _bc.encode(self.astuple())
        return x

    @classmethod
    def decode(cls, enc: bytes) -> Self:
        return cls.fromtuple(_bc.decode(enc))


@dataclass(frozen=True)
class PendingReturn(TaggedTuple):
    tag = 1
    root_id: str
    call_hash: str
    topic: str


@dataclass(frozen=True)
class ScheduleMessage(TaggedTupleStrings):
    tag = 2
    root_id: str
    call_hash: str
